### ISSUE TITLE
move PMF code to its own file and add other PMF methods (may require reorganizing class).

### ISSUE THREAD

**Issue by mrshirts** [choderalab/pymbar#145]
I'm thinking of adding some more PMF tools (for example, based on minimizing KL divergence) and am wondering if we want to move the PMF tools out of mbar.py.  Since they are currently methods of the class, it would require some reorganization.  It may not actually make sense, but I though I'd check.

**Comment by kyleabeauchamp** [comment]
I don't have a good sense of how this would look yet.  However, I do like the idea of smaller objects.  For example, in pymbar2 I felt like it was hard to keep track of everything that an `MBAR` object could do.

**Comment by jchodera** [comment]
Moving PMF tools into a separate helper class sounds like a great idea.

To help us figure out what this would look like, how about suggesting a few use cases?

**Comment by mrshirts** [comment]
use cases for different PMF calculations, or use cases for smaller objects?

**Comment by jchodera** [comment]
> use cases for different PMF calculations, or use cases for smaller objects?

Smaller objects. 

For example, we currently have things that look like this:
```python
from pymbar import MBAR
mbar = MBAR(u_kln, N_k, method='adaptive')
# ... compute binning ourselves...
[f_i, df_i] = mbar.computePMF(u_kn, bin_kn, nbins, uncertainties='from-specified', pmf_reference = zeroindex)  
```
and instead we might want to roll some of the machinery from the examples into a `pmftools` class, like this:
```python
from pymbar import MBAR, pmftools
mbar = MBAR(u_kln, N_k, method='adaptive')
# Bin data (x_kn) into non-overlapping histogram bins in 1D using adjustable-width bins.
[xcenter_i, bin_kn, xleft_i, xright_i, bin_widths] = pmftools.bin_histogram_1d(x_kn, nbins, method='fixed-number')
# Compute free energy estimate of bins.
[f_i, df_i] = pmftools.compute_categorical_pmf(bin_kn, nbins, uncertainties='from-specified', pmf_reference = zeroindex, bin_volumes=bin_widths)  
```
or
```python
from pymbar import MBAR, pmftools
mbar = MBAR(u_kln, N_k, method='adaptive')
# Bin data (x_kn) into non-overlapping histogram bins in 1D using adjustable-width bins.
[xcenter_i, bin_kn, xleft_i, xright_i] = pmftools.bin_histogram_1d(x_kn, nbins, method='fixed-width-eliminate-empty')
# Compute free energy estimate of bins.
[f_i, df_i] = pmftools.compute_categorical_pmf(bin_kn, nbins, uncertainties='from-specified', pmf_reference = zeroindex)  
```

**Comment by mrshirts** [comment]
Kyle, do you want to take a pass at just converting the basic PMF function into a pmf_tools?  I think you have a better grasp and the python organization.  I can convert examples after.  

This will also make a good place for me to put in the variant of the York PMF calculations.  

BTW, I think this is lower priority to the MBAR acceleration paper draft.

**Comment by mrshirts** [comment]
I might start this so I can get some of the alternate PMF code into the repository.  

Is passing the mbar object into a pmf_tools the best way to get the information into there, or is there some other way involving inheritance?  I didn't see a clear way in the code above.

**Comment by kyleabeauchamp** [comment]
IMHO, if we only require a few pieces of information, we just pass data around as arrays.  I'd avoid passing objects around unless it's really needed to hide complexity.  My justification for this is that people will ideally want to mix and match different MBAR / PMF estimators.  

Again, it's hard to say until we have a better idea of what all the necessary ingredients are.  

Also, we should decide whether this is a pymbar3.0 target or not.  If it's not, then we might want to holds off on making big changes in favor of clearing out all the 3.0 tagged issues, which are mainly documentation at this point IMHO.

**Comment by mrshirts** [comment]
> IMHO, if we only require a few pieces of information, we just pass data around as arrays. I'd avoid passing objects around unless it's really needed to hide complexity. My justification for this is that people will ideally want to mix and match different MBAR / PMF estimators.

Let's see.  Figuring out exactly what to extract can get complicated.

Current plan -- add new methods in a new branch, and then see after how to best organize.

**Comment by mrshirts** [comment]
Addressed in #327
